### ISSUE TITLE
Proposal: automate post installation steps for v3

### ISSUE THREAD

**Issue by M365Bass** [pnp/pnpjs#2184]
### Category
- [x] Enhancement
- [ ] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [Tested on v3.1.0 but should work with all v3]
Please specify what version(s) of SharePoint you are targeting: [NA]

### ~~Expected / Desired Behavior / Question~~ Proposal
**TL;DR** automate the manual steps after installing PnP JS using a node script triggered via npm postinstall.

In a recent project, I installed PnP JS v3 and went through the steps [here](https://pnp.github.io/pnpjs/getting-started/#install). While the steps are very detailed and sufficient, it still took some time to be able to get everything in order. I remembered that PnP property controls had a similar setup where the PropertyControlStrings was added automatically after setup. I used that as an example and started to build a similar script to automate the following:

1. Replace `rush-stack-compiler` version in `package.json`
2. Update `extends` prop in `tsconfig`
3. Disable `tslint` in `gulpfile.js`
4. [Needs feedback] Add `spfi, SPFx imports` to any WebParts as well as update `onInit`
From there, you'd only need to run npm install once more to replace the `rush-stack-compiler` (which we could look at automating as well).

I'm looking for feedback on 1) whether this is a feature we'd like to add to simplify the installation process and 2) whether the last item in the above list is of use given that it may be a bit intrusive to affect all WebParts, especially in big projects not to mention the added complexity with newer spfx versions as well as other project types (extensions, libraries, cards & whatever comes next).

I wanted to get feedback early to see whether this is of interest and though some code would help explain things. I'll submit a PR related to this ticket with some draft code. As the script uses npm postinstall, it would run automatically after installing the package. The script would require some refactoring if the idea is of interest (which I'm happy to do). 

Thanks!

**Comment by patrick-rodgers** [comment]
Hi @M365Bass - I think this is a super cool idea! Is there a way to do it such that it isn't automatically run? I'd like folks to opt in vs opt out.

**Comment by juliemturner** [comment]
I agree, although I love this for most users to simplify the process, I would not want this for myself... that said it's a great idea.

**Comment by M365Bass** [comment]
Thank you both for the swift feedback. Do you think all the steps should be optional or just the last step affecting WebParts?

Given the first three steps are mandatory config steps for the library to function, I thought these would run automatically post-installation. I'm intrigued to understand if there are more complex setups where making these changes automatically rather than manually would make sense.

**Comment by juliemturner** [comment]
The primary issue as I see it is that the first three steps aren't for all types of projects that use the library, only SPFx....

**Comment by M365Bass** [comment]
Great points. What if I check the current project for known SPFx dependencies like sp-core-library, sp-webpart-base and only make the changes if it is indeed an SPFx project?

For rush stack compiler, we can have a config file with a version for each SPFx version (so SPFx 1.13 and 1.14 changes rush stack compiler from 3.9 to 4.2 and you can add a similar config for 1.15 when it comes out)

**Comment by patrick-rodgers** [comment]
@waldekmastykarz - wonder if this is better implemented as a cli command? I know you already do some other project upgrade work, so perhaps we could model on that work?

This is a great idea just want to find the best place to make it easy for people without forcing project changes on folks.

**Comment by waldekmastykarz** [comment]
We could definitely introduce a command, like `spfx project pnpjs add` which would analyze the project, detect its version and provide the recommended steps to update the code. Right now, it would be a report, similar to what you get for upgrading an SPFx project or validating its config. If in the future we introduce automatically applying the necessary changes, then we'd have the kind of automation I believe you mentioned @M365Bass.

If you think this approach would be valuable enough for developers, then let's create an issue in the CLI for M365 repo and continue the discussion there.

**Comment by M365Bass** [comment]
While the M365 CLI is a an awesome tool, I would see a CLI command as an overhead as it would require installing another package to automate the configuration steps. There's also the aspect of using code tour as opposed to making the actual changes.

I mentioned earlier that I can address the earlier concerns by checking if the component installing PnPJS is an SPFx component and only making changes then. This would ensure that only SPFx component are affected without touching other types of projects. Making things configurable to support future SPFx & Rush stack compiler versions is definitely possible as well.

I started this as a proposal as I understand it is a relatively core change and may not align with your future direction & roadmap. I'm just keen to streamline the experience for new SPFx developers by removing any manual configuration.

Cheers

**Comment by juliemturner** [comment]
While we love your idea on principle, the problem is it's too opinionated which doesn't really fit in with our ethos. Further, if all things go as planned, the first 3 steps will be unneeded for SPFx 1.15 (and beyond). That said I really do love the idea of a "getting started" script I just don't know where it can go. I think the CLI idea is inspired as that's where these types of scenarios make sense... and could be expanded to include a pnpjsConfig.ts file... again, great for those getting started, not necessarily what all users would want....

**Comment by M365Bass** [comment]
Thanks @juliemturner, I appreciate your feedback. I think given how it would resolve on its own with the next release of SPFx, there's no need for the first three steps to be implemented. If there's a need for further config automation - whether as a post installation script or a command in M365 CLI - I'm happy to pick up the work once we identify the items to automate.

Thanks,
B

**Comment by juliemturner** [comment]
I will discuss it again with @patrick-rodgers -- 
~~I just typed a whole bunch of stuff and then erased it when I got to the point where I was saying that even the most season developers use templates and suggesting that I have a bunch of VSCode snippets that I use to make my life simpler for this boilerplate type code...~~

Doesn't negate the idea that it could be a script and that that script could live in the CLI but feels like another alternative answer. So, I'm not sure I have an answer for where it should go, inside the library is the wrong answer -- but there's a potential argument for a PnP VSCode Snippets repo. What would concern me is that something like this would end up getting bogged down and cluttered with too many submissions, effectively negating its usefulness. What I'm suggesting shouldn't be a code sample gallery we have those... a lot of those. Maybe it's just a repo with a snippets file per PnP SDK. So, we'd take submissions, but they would be modifications of one file. Long story short, I'm not sure, I need to think more and will discuss with the wider team.
